executor/memtable_reader.go

```diff
@@ -698,3 +700,507 @@ func (e *clusterLogRetriever) close() error {
+		if err != nil {
+			return nil, err
+		}
+		if row != nil {
+			finalRows = append(finalRows, row)
+		}
+		minTimeItem.messages.HistoryHotRegion = minTimeItem.messages.HistoryHotRegion[1:]
+		// Fetch next message item
+		if len(minTimeItem.messages.HistoryHotRegion) != 0 {
+			heap.Push(e.heap, minTimeItem)
+		}
+	}
+	// All streams are drained
+	e.isDrained = e.heap.Len() == 0
+	return finalRows, nil
+}
+
+func (e *hotRegionsHistoryRetriver) getHotRegionRowWithSchemaInfo(
+	hisHotRegion *HistoryHotRegion,
+	tikvHelper *helper.Helper,
+	allSchemas []*model.DBInfo,
+	tz *time.Location,
+) ([]types.Datum, error) {
+	_, startKey, _ := codec.DecodeBytes(hisHotRegion.StartKey, []byte{})
+	_, endKey, _ := codec.DecodeBytes(hisHotRegion.EndKey, []byte{})
+	region := &tikv.KeyLocation{StartKey: startKey, EndKey: endKey}
+	hotRange, err := helper.NewRegionFrameRange(region)
+	if err != nil {
+		return nil, err
+	}
+
+	f := tikvHelper.FindTableIndexOfRegion(allSchemas, hotRange)
+	// Ignore row without corresponding schema f.
+	if f == nil {
+		return nil, nil
+	}
+	row := make([]types.Datum, len(infoschema.TableTiDBHotRegionsHistoryCols))
+	updateTimestamp := time.Unix(hisHotRegion.UpdateTime/1000, (hisHotRegion.UpdateTime%1000)*int64(time.Millisecond))
+
+	if updateTimestamp.Location() != tz {
+		updateTimestamp.In(tz)
+	}
+	updateTime := types.NewTime(types.FromGoTime(updateTimestamp), mysql.TypeTimestamp, types.MinFsp)
+	row[0].SetMysqlTime(updateTime)
+	row[1].SetString(strings.ToUpper(f.DBName), mysql.DefaultCollationName)
+	row[2].SetString(strings.ToUpper(f.TableName), mysql.DefaultCollationName)
+	row[3].SetInt64(f.TableID)
+	if f.IndexName != "" {
+		row[4].SetString(strings.ToUpper(f.IndexName), mysql.DefaultCollationName)
+		row[5].SetInt64(f.IndexID)
+	} else {
+		row[4].SetNull()
+		row[5].SetNull()
+	}
+	row[6].SetInt64(int64(hisHotRegion.RegionID))
+	row[7].SetInt64(int64(hisHotRegion.StoreID))
+	row[8].SetInt64(int64(hisHotRegion.PeerID))
+	if hisHotRegion.IsLearner {
+		row[9].SetInt64(1)
+	} else {
+		row[9].SetInt64(0)
+	}
+	if hisHotRegion.IsLeader {
+		row[10].SetInt64(1)
+	} else {
+		row[10].SetInt64(0)
+	}
+
+	row[11].SetString(strings.ToUpper(hisHotRegion.HotRegionType), mysql.DefaultCollationName)
+	if hisHotRegion.HotDegree != 0 {
+		row[12].SetInt64(hisHotRegion.HotDegree)
+	} else {
+		row[12].SetNull()
+	}
+	if hisHotRegion.FlowBytes != 0 {
+		row[13].SetFloat64(float64(hisHotRegion.FlowBytes))
+	} else {
+		row[13].SetNull()
+	}
+	if hisHotRegion.KeyRate != 0 {
+		row[14].SetFloat64(float64(hisHotRegion.KeyRate))
+	} else {
+		row[14].SetNull()
+	}
+	if hisHotRegion.QueryRate != 0 {
+		row[15].SetFloat64(float64(hisHotRegion.QueryRate))
+	} else {
+		row[15].SetNull()
+	}
+	return row, nil
+}
+
+func (e *hotRegionsHistoryRetriver) close() error {
+	if e.cancel != nil {
+		e.cancel()
+	}
+	return nil
+}
+
+func (e *hotRegionsHistoryRetriver) getRuntimeStats() execdetails.RuntimeStats {
+	return nil
+}
```

Comment:
it's work of previous pr, it should be add by `update branch` after previous pr merged.